File: spoj/1436-PT07Y/solution_1436.py
# Time limit exceed
def is_graph_dfs(edges, nodes):
    stack = []
    visitedNodes = [False for x in range(nodes + 1)]

    stack.append(edges[0][0])
    visitedNodes[edges[0][0]] = True
    visitedCount = 1

    while visitedCount < nodes and len(stack) != 0:
        for first, second in edges:
            if stack[-1] == first and not visitedNodes[second]:
                visitedCount += 1
                visitedNodes[second] = True
                stack.append(second)
        stack.pop()

    return visitedCount == nodes


if __name__ == "__main__":
    nodes, e = map(int, input().split(" "))
    edges = []
    for x in range(e):
        edges.append(list(map(int, input().split(" "))))

    if is_graph_dfs(edges, nodes) and (nodes - 1) == e:
        print("YES")
    else:
        print("NO")
# 5 4
# 1 2
# 1 3
# 3 4
# 4 5

File: spoj/1436-PT07Y/solution_1436.cpp
#include <bits/stdc++.h>

using namespace std;

int main()
{
    int n, e;
    cin >> n >> e;

    int arr[e][2];
    int stack[n], top = -1;

    bool visitedNodes[n];
    memset(visitedNodes, false, n);

    for (int i = 0; i < e; i++)
    {
        for (int j = 0; j < 2; j++)
        {
            cin >> arr[i][j];
        }
    }

    stack[++top] = arr[0][0];
    visitedNodes[stack[top] - 1] = true;
    int visited = 1;

    while (visited < n && top != -1)
    {
        for (int i = 0; i < e; i++)
        {
            int first = arr[i][0];
            int second = arr[i][1];
            bool isNotVisited = !visitedNodes[second - 1];
            if (first == stack[top] && isNotVisited)
            {
                visited++;
                stack[++top] = second;
                visitedNodes[second - 1] = true;
            }
        }
        top--;
    }

    if ((n - 1) == e && visited == n)
        cout << "YES";
    else
        cout << "NO";
    return 0;
}